mithril-aggregator/src/services/prover.rs

```diff
@@ -114,38 +109,51 @@ impl ProverService for MithrilProverService {
         up_to: &CardanoDbBeacon,
         transaction_hashes: &[TransactionHash],
     ) -> StdResult<Vec<CardanoTransactionsSetProof>> {
-        // 1 - Get transactions to prove per block range
-        let transactions_to_prove = self
-            .get_transactions_by_hashes_with_block_range(transaction_hashes.to_vec())
+        // 1 - Compute the set of block ranges with transactions to prove
+        let block_ranges_transactions = self.get_block_ranges(transaction_hashes).await?;
+        let block_range_transactions = self
+            .get_all_transactions_for_block_ranges(&block_ranges_transactions)
             .await?;
 
-        // 2 - Compute Transactions Merkle Tree
-        let transactions = self.transaction_retriever.get_up_to(up_to).await?;
-        let mk_map = self.compute_merkle_map_from_transactions(transactions)?;
-
-        // 3 - Compute proof for each transaction to prove
-        let mut transaction_hashes_certified = vec![];
-        for (_block_range, transaction) in transactions_to_prove {
-            let mk_tree_node_transaction_hash: MKTreeNode =
-                transaction.transaction_hash.to_owned().into();
-            if mk_map
-                .compute_proof(&[mk_tree_node_transaction_hash])
-                .is_ok()
-            {
-                transaction_hashes_certified.push(transaction.transaction_hash);
-            }
+        // 2 - Compute block ranges sub Merkle trees
+        let mut mk_trees = BTreeMap::new();
+        for (block_range, transactions) in block_range_transactions {
+            let mk_tree = MKTree::new(
+                &transactions
+                    .iter()
+                    .map(|t| t.transaction_hash.clone())
+                    .collect::<Vec<_>>(),
+            )?;
```

Comment:
If I'm not wrong, since `CardanoTransaction` can be converted into a `MKTreeNode`, you can simplify this to:
```suggestion
            let mk_tree = MKTree::new(transactions)?;
```